docs/docs/render-props.md

```diff
@@ -0,0 +1,224 @@
+
+For example, let's say we have a `<Cat>` component that renders the image of a cat chasing the mouse around the screen. We might use a `<Cat mouse={{ x, y }}>` prop to tell the component the coordinates of the mouse so it knows where to position the image on the screen.
+
+As a first pass, you might try rendering the `<Cat>` *inside `<Mouse>`'s `render` method*, like this:
+
+```js
+class Cat extends React.Component {
+  render() {
+    const mouse = this.props.mouse
+    return (
+      <img src="/cat.jpg" style={{ position: 'absolute', left: mouse.x, top: mouse.y }}/>
+    );
+  }
+}
+
+class MouseWithCat extends React.Component {
+  constructor(props) {
+    super(props);
+    this.handleMouseMove = this.handleMouseMove.bind(this);
+    this.state = { x: 0, y: 0 };
+  }
+
+  handleMouseMove(event) {
+    this.setState({
+      x: event.clientX,
+      y: event.clientY
+    });
+  }
+
+  render() {
+    return (
+      <div style={{ height: '100%' }} onMouseMove={this.handleMouseMove}>
+
+        {/*
+          We could just swap out the <p> for a <Cat> here ... but then
+          we would need to create a separate <MouseWithSomethingElse>
+          component every time we need to use it, so <MouseWithCat>
+          isn't really reusable yet.
+        */}
+        <Cat mouse={this.state}/>
+      </div>
+    );
+  }
+}
+
+class MouseTracker extends React.Component {
+  render() {
+    return (
+      <div>
+        <h1>Move the mouse around!</h1>
+        <MouseWithCat/>
+      </div>
+    );
+  }
+}
+```
+
+This approach will work for our specific use case, but we haven't achieved the objective of truly encapsulating the behavior in a reusable way. Now, every time we want the mouse position for a different use case, we have to create a new component (i.e. essentially another `<MouseWithCat>`) that renders something specifically for that use case.
+
+Here's where the render prop comes in: Instead of hard-coding a `<Cat>` inside a `<Mouse>` component, and effectively changing its rendered output, we can provide `<Mouse>` with a function prop that it uses to dynamically determine what to render–a render prop.
+
+```js
+class Cat extends React.Component {
+  render() {
+    const mouse = this.props.mouse
+    return (
+      <img src="/cat.jpg" style={{ position: 'absolute', left: mouse.x, top: mouse.y }}/>
+    );
+  }
+}
+
+class Mouse extends React.Component {
+  constructor(props) {
+    super(props);
+    this.handleMouseMove = this.handleMouseMove.bind(this);
+    this.state = { x: 0, y: 0 };
+  }
+
+  handleMouseMove(event) {
+    this.setState({
+      x: event.clientX,
+      y: event.clientY
+    });
+  }
+
+  render() {
+    return (
+      <div style={{ height: '100%' }} onMouseMove={this.handleMouseMove}>
+
+        {/*
+          Instead of providing a static representation of what <Mouse> renders,
+          use the `render` prop to dynamically determine what to render.
+        */}
+        {this.props.render(this.state)}
```

Comment:
First, for security reasons, you probably shouldn't expose `state` directly to a foreign component. You can't be guaranteed that they won't mutate it.

Second, why not use JSX in this situation? You're expecting the caller to pass in a component, so why not just render it with JSX? You did so on line 126 above, so why change how `Cat` is rendered just because the implementation of `Cat` is passed in as a prop?

With that, I'd like to see `render` like this. Note how I spread `state`.

```jsx
render() {
  const Cat = this.props.render;
  return (
    <div style={{ height: '100%' }} onMouseMove={this.handleMouseMove}>
      <Cat {...state} />
    </div>
  );
}
```

In my article for AmericanExpress.io titled ["Function as Child Components Are an Anti-Pattern"](http://americanexpress.io/faccs-are-an-antipattern/), I talk about various implimentation of **render callback**.  I call this specific pattern **Component Injection** as you are injecting a component. I find myself using this pattern frequently.

Reply:
Calling a function like this `this.props.render()` and this `<this.props.render/>` are extremely different. One nice thing about render props is that you can inline behavior. If you use `createElement` on an inline function you're going to get mounting/unmounting. This is why Route in React Router has a `render` prop (called as a function) and a `component` prop (called with `createElement`). Not the same thing.

Additionally, the point is to compose Cat and Mouse w/o either having to know about eachother.

Reply:
As for mutating the state object, the only risks are to the application. This component doesn't use state anywhere other than to pass it down in render. So, whether we give them a copy of state or the actual object doesn't change anything. They'll mutate the copy and get the same bugs.

Reply:
@ryanflorence,
 
See my implementation of **MouseTracker** in this [CodeSandbox](https://codesandbox.io/s/w0r04vryqk).

> If you use createElement on an inline function you're going to get mounting/unmounting on every render, which you never want.

If `Cat` is mounted/umounted on each render (as you suggested), then why is the console output in `componentWillMount`  displayed only once? Note that I've only turned `Cat` into a class component so that I could `console.log` the mount lifecycle. And wouldn't the same thing hold true in line 192? 

> Additionally, the point is to compose Cat and Mouse w/o either having to know about the other, so making `renderCat` obliterates the decoupled composition of render props.

I'll give you that one and I completely agree that one shouldn't know about the other. I was being a bit to literal and following @mjackson's prior naming.

Reply:
@donavon it comes down to whether the render component is defined inline or not--React reconciles based on the type of the component. In your demo Cat it is static so a createElement call preserves the instance. It seems the RR folks really want to encourage a pattern that allows for inline definition.

Reply:
@alexkrolick,

Ahh... Now I get what @ryanflorence was saying. Thx for clarifying.

Reply:
@alexkrolick 

> It seems the RR folks really want to encourage a pattern that allows for inline definition.

I know that my next statement is like throwing gasoline on a slow burning Twitter controversy (and even cause SOME TO SCREAM), but... Inline component definitions (like inline labmda function handlers in `render`) are just a bad idea from a memory management/garbage collection perspective. I think we should _discourage_ the pattern that allows for inline definition.

Reply:
>  inline labmda function handlers in render) are just a bad idea from a memory management/garbage collection perspective.

I agree with this, and actually wrote a lint rule for it (https://github.com/yannickcr/eslint-plugin-react/commit/19faaac2f78c6568635165e26d9a05edabcc01fc#diff-3bb8bcac6ca99e0bf2c936111e93256e). It can have nontrivial effect on performance if it's done on a hot code path in a large app. My benchmarking was done a few years ago though, so perhaps newer browsers are better?

Reply:
Since this is only really an issue in large apps, perhaps I could just put a note here that says something like

> If a render prop returns a large component tree, it's a good idea to use an instance method if you can instead of writing the function inline in your JSX.

What do you think?

Reply:
Created a CodePen to demo which patterns definitely cause a repaint in the DOM:

https://codepen.io/alexkrolick/pen/WZwMYW?editors=0010

![kapture 2017-09-20 at 0 29 53](https://user-images.githubusercontent.com/1571667/30631908-f044d696-9d9a-11e7-8582-1ab3dfcbb8c3.gif)

Note that render props that are called as functions rather than components (createElement/JSX) are the only inline-able ones that doesn't cause a redraw. 

This is totally aside from any memory issues that arise from defining functions in render.

Reply:
@alexkrolick It also useful to add another `Inline`, but this time a class:

```
const InlineClass = class InlineClass extends Component {
    render() {
        return <span>{this.props.foo}</span>
    }
}
```

Reply:
@alexkrolick Very nice. It demonstrates exactly what's going on. Thx for putting this together! I'd like to point out your last comment once again.

> **This is totally aside from any memory issues that arise from defining functions in render.**